### PR TITLE
feat(daemon): containerized dependency management

### PR DESCRIPTION
**Pull request by deepankarm**
## JinaD 2.0

The primary goal of this item is to containerize every object creation under JinaD, thereby separating the environment for each object, allowing custom build & non-conflicting networks. [Usage of JinaD is documented here.](https://github.com/jina-ai/jina/blob/daemon-2.0/.github/2.0/cookbooks/Daemon.md)

Items achieved under this PR:

### Store:

- [x] Refactor Workspace Store & enable docker-build.
- [x] Define a Container Store & enable docker-run.
- [x] Enable (de)serialization with stores.
- [ ] Add thread locking to store object
- [ ] Add validation to the store metadata.
- [ ] Optionally enable building store from all local images/containers.

###### Runtime:

- [x] Enable default bash runtime with fluentd. 
- [x] Add build context for actual run & devel. 
- [x] Enable a different runtime with jinad jr (another fastapi layer) inside docker.
- [x] Setting Env vars during run

###### Network:

- [x] Create docker network during workspace creation.
- [x] Handle port-mapping during docker run ?? 

### Configuration:

- [x] Define a .jinad syntax (possibly ini) & create dockerfile from it.
- [x] requirements.txt
- [ ] setup commands
- [x] entrypoint command for projects

### Request Handling:

- [x] Build web-queue-worker architecture.
- [x] Enable endpoint for tracking requests

### Endpoints:

- [x] Ensure logging works inside containers.
- [x] Handle `DELETE` methods.
- [x] Add `Depends` to endpoints to handle params/validation etc.
- [ ] Support environment variables for all endpoints.
- [x] Separate POST vs PUT for all endpoints.
- [x] Enable endpoint for `/projects` ?? (tobi)
- [ ] Enable file reading endpoint inside workspaces. Store file metadata.
- [ ] Add a `/stats` endpoint to execute docker stats (Tobi)

### Housekeeping:

- [x] Add custom-type (DaemonID) for validation str+uuid

### Bugs:

- [x] DELETE flows during DELETE workspaces
- [x] Logging during docker run/build
- [x] Logging connection doesn't drop even when client has dropped
- [x] Change log endpoint to accept only id (fetch workspace_id from id if required)

##### Pending items:

- Currently, JinaD builds all images from `test-daemon` which is only available during development, Once this PR is merged, we'll add support to pulling from Docker Hub in a new PR. 
- Except `test_simple_hub_pod` & `rolling_update`, all tests under `distributed` have been skipped for now. Most of the topologies are covered under `test_single_instance` and few of the distributed tests would be added back in next PRs.
- Unchecked items will be achieved in the next PRs.

### CODE REVIEW

File: daemon/api/dependencies.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    def __init__(self, workspace_id: DaemonID, filename: str) -> None:
+        self.workspace_id = workspace_id
+        self.filename = filename
+        self.localpath = self.validate()
+        self.id = DaemonID('jflow')
+        self.params = FlowModel(
+            uses=self.filename, workspace_id=self.workspace_id.jid, identity=self.id
+        )
+
+    @cached_property
```
Comment (JoanFM) [inline daemon/api/dependencies.py:28]: why does it need to be a cached property? can't it be simply a `property`? or can we just cache it ourselves? I think the extra work done by `cached_property` is unwanted

File: daemon/api/dependencies.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        `port_expose` for gateway needs to be set before starting the container.
+        Flow yaml might have `port_expose` set already, if not, set it to random_port
+        """
+        f = Flow.load_config(str(self.localpath))
+        _port_expose = f._common_kwargs.get('port_expose')
+        if not _port_expose:
+            _port_expose = random_port()
+        return _port_expose
+
+    @cached_property
```
Comment (JoanFM) [inline daemon/api/dependencies.py:40]: same here

File: daemon/api/dependencies.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    def ports(self) -> Dict[str, str]:
+        return {f'{self.port_expose}/tcp': self.port_expose}
```
Comment (JoanFM) [inline daemon/api/dependencies.py:42]: it is weird to need this `property` right? a single valued dictionary?

Comment (JoanFM) [reply daemon/api/dependencies.py:42]: Seeing the other objects, maybe u need to build an abstract interface?

File: daemon/dockerize.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+            },
+            # TODO: without adding slash, it fails on WSL (needs to checked on linux/mac)
+            '//var/run/docker.sock': {'bind': '/var/run/docker.sock'},
+        }
+
+    @classmethod
+    def environment(cls) -> Dict[str, str]:
+        return {'JINA_LOG_WORKSPACE': '/workspace/logs', 'JINA_RANDOM_PORTS': 'True'}
+
+    def remove(cls, id: DaemonID):
```
Comment (JoanFM) [inline daemon/dockerize.py:231]: not a `class method`?

File: daemon/dockerize.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+
+
+if TYPE_CHECKING:
+    from .files import DaemonFile
+    from docker.models.networks import Network
+    from docker.models.containers import Container
+    from docker.client import APIClient, DockerClient
+
+
+class Dockerizer:
```
Comment (JoanFM) [inline daemon/dockerize.py:38]: Is this a common `singleton` pattern in `python`? Asking out of ignorance

File: daemon/stores/__init__.py [hunk 6]
```diff
@@ -1,9 +1,63 @@
...
+from .workspaces import WorkspaceStore
+from ..models import DaemonID
+from ..models.enums import IDLiterals
+
+if TYPE_CHECKING:
+    from .partial import PartialStore
+
+
+def _get_store(kind: str):
+    if kind == 'pea':
```
Comment (JoanFM) [inline daemon/stores/__init__.py:17]: having a `static` map may be more elegant

File: daemon/api/dependencies.py (hunk 1 above)
Comment (deepankarm) [reply daemon/api/dependencies.py]: I've been using `cached_property` as a setter & a getter. We'll be moving away from `cached_property` wherever not required in the next set of cleanup PRs.

**Review by jacobowitz — DISMISSED** [review]
from my side this is good to go :)

File: tests/daemon/unit/api/endpoints/test_logs.py [hunk 7]
```diff
@@ -13,50 +17,69 @@
...
+        arguments=ContainerArguments(command='command', object=Flow()),
+        workspace_id=workspace_id,
+    )
 
-def test_logs_invalid_flow(fastapi_client):
-    response = fastapi_client.get(f'/logs/{workspace_id}/{nonexisting_id}')
-    assert response.status_code == 404
+
+def setup_module():
+    print('setup', get_workspace_path(workspace_id, flow_id))
```
Comment (maximilianwerk) [inline tests/daemon/unit/api/endpoints/test_logs.py:60]: remove leftover print?

File: tests/daemon/unit/api/endpoints/partial/test_flow.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+    flow_model = FlowModel()
+    flow_model.uses = f'{cur_dir}/good_flow_dummy.yml'
+    response = partial_flow_client.post(
+        api, json=flow_model.dict(exclude={'log_config'}), params={'port_expose': 56789}
+    )
+    assert response
+
+    response = partial_flow_client.get(api)
+    assert response
+    print(response.json())
```
Comment (maximilianwerk) [inline tests/daemon/unit/api/endpoints/partial/test_flow.py:21]: remove `print`

File: jina/enums.py [hunk 9]
```diff
@@ -429,6 +429,17 @@ def is_scipy_stackable(self) -> bool:
         return self.value in [1, 2]
 
 
+class RemoteWorkspaceState(str, Enum):
```
Comment (hanxiao) [inline jina/enums.py:432]: simply use `BetterEnum`?

File: jina/flow/base.py [hunk 10]
```diff
@@ -1558,6 +1558,28 @@ def rolling_update(self, pod_name: str, dump_path: Optional[str] = None):
...
+        :param shards: nr of shards to dump for
+        :param on_done: the function to be called after the dump request has completed
+        """
+        self.post(
+            on='/dump',
+            target_peapod=pod_name,
+            parameters={'dump_path': dump_path, 'shards': shards},
+            on_done=on_done,
+        )
+
```
Comment (hanxiao) [inline jina/flow/base.py:1582]: delete, no sugary syntax on `post` is allowed here. We do not make assumption anymore on executors' endpoints
```suggestion

```

File: jina/peapods/pods/__init__.py [hunk 11]
```diff
@@ -119,6 +121,38 @@ def close(self):
...
+
+        def valid_path(path):
+            try:
+                complete_path(path)
+                return True
+            except FileNotFoundError:
+                return False
+
+        _upload_files = set()
+        for param in ['uses', 'uses_internal', 'uses_before', 'uses_after']:
```
Comment (hanxiao) [inline jina/peapods/pods/__init__.py:137]: is `uses_internal` still valid? @JoanFM

File: jina/peapods/runtimes/jinad/__init__.py [hunk 12]
```diff
@@ -17,26 +19,26 @@ def __init__(self, args: 'argparse.Namespace', ctrl_addr: str):
...
         self.port_expose = args.port_expose
-        self.api = PeaDaemonClient(
+        # TODO: args.timeout_ready is always set to -1 for JinadRuntime so that wait_for_success doesn't fail in Pea,
+        # so it can't be used for Client timeout. Setting `timeout` to `None` would wait forever in `DaemonClient`,
+        # which is not ideal. Setting it to 15 secs for testing.
+        self.workspace_api = WorkspaceDaemonClient(
+            host=self.host,
+            port=self.port_expose,
+            logger=self.logger,
+            timeout=15,
```
Comment (hanxiao) [inline jina/peapods/runtimes/jinad/__init__.py:29]: `15` looks like a leftover value from debugging?

File: jina/peapods/runtimes/jinad/__init__.py [hunk 13]
```diff
@@ -17,26 +19,26 @@ def __init__(self, args: 'argparse.Namespace', ctrl_addr: str):
...
+        )
+        self.pea_api = PeaDaemonClient(
             host=self.host,
             port=self.port_expose,
             logger=self.logger,
-            timeout=self.args.timeout_ready,
+            timeout=15,
```
Comment (hanxiao) [inline jina/peapods/runtimes/jinad/__init__.py:35]: same here

File: jina/peapods/runtimes/jinad/client.py [hunk 14]
```diff
@@ -245,6 +139,23 @@ def _mask_args(self, args: 'argparse.Namespace'):
...
+            return field
+
+        for f in ('uses', 'uses_after', 'uses_before', 'uses_internal', 'py_modules'):
+            attr = getattr(_args, f, None)
+            if not attr:
+                continue
+            setattr(_args, f, [basename(m) for m in attr]) if isinstance(
+                attr, list
+            ) else setattr(_args, f, basename(attr))
+
```
Comment (hanxiao) [inline jina/peapods/runtimes/jinad/client.py:158]: i think this part of the code especially the `basename` function has to be aligned with Hubble, in order to use Hubble images. Currently hubble's schemes are `jinahub://` or `jinahub+docker://`

File: jina/flow/base.py [hunk 15]
```diff
@@ -1558,6 +1558,28 @@ def rolling_update(self, pod_name: str, dump_path: Optional[str] = None):
...
+        :param pod_name: pod to target with the dump request
+        :param dump_path: the path to which to dump
+        :param shards: nr of shards to dump for
+        :param on_done: the function to be called after the dump request has completed
+        """
+        self.post(
+            on='/dump',
+            target_peapod=pod_name,
+            parameters={'dump_path': dump_path, 'shards': shards},
```
Comment (maximilianwerk) [inline jina/flow/base.py:1579]: I think there needs to be a way to specify more dynamic argument here as well. Imagine parameters like `dump_format`. I am good either way: add a `**kwargs` to the function definition and then passed to the `parameters` dict or a `parameters` argument which is altered with `dump_path` and `shards` afterwards.

File: jina/peapods/peas/__init__.py [hunk 16]
```diff
@@ -310,6 +310,9 @@ def _get_runtime_cls(self) -> Tuple[Any, bool]:
         is_remote_controlled = False
         if self.args.host != __default_host__:
             self.args.runtime_cls = 'JinadRuntime'
+            # NOTE: remote pea would also create a remote workspace which might take alot of time.
+            # setting it to -1 so that wait_start_success doesn't fail
+            self.args.timeout_ready = -1
             is_remote_controlled = True
         if self.args.runtime_cls == 'ZEDRuntime' and self.args.uses.startswith(
             'docker://'
```
Comment (hanxiao) [inline jina/peapods/peas/__init__.py:318]: this may need a change.

File: jina/peapods/runtimes/jinad/client.py (hunk 14 above)
Comment (hanxiao) [reply jina/peapods/runtimes/jinad/client.py:158]: @numb3r3

File: jina/parsers/peapods/base.py [hunk 17]
```diff
@@ -69,3 +69,12 @@ def mixin_base_ppr_parser(parser, with_identity: bool = True):
...
             else argparse.SUPPRESS,
         )
+
+    gp.add_argument(
+        '--workspace-id',
+        type=str,
+        default=random_identity(),
+        help='the UUID for identifying the workspace. When not given a random id will be assigned.'
+        'Multiple Pea/Pod/Flow will work under the same workspace if they share the same '
+        '`workspace-id`.',
```
Comment (hanxiao) [inline jina/parsers/peapods/base.py:79]:
```suggestion
        '`workspace-id`.' if _SHOW_ALL_ARGS else argparse.SUPPRESS,
```
don't think user need to know about this, let's hide it from CLI

File: jina/peapods/pods/__init__.py [hunk 18]
```diff
@@ -119,6 +121,38 @@ def close(self):
...
+            param_value = getattr(args, param, None)
+            if param_value and valid_path(param_value):
+                _upload_files.add(param_value)
+
+        if getattr(args, 'py_modules', None):
```
Comment (maximilianwerk) [inline jina/peapods/pods/__init__.py:142]: I'd prefer `if hasattr(args, 'py_modules'):`

File: jina/peapods/runtimes/container/__init__.py [hunk 19]
```diff
@@ -29,6 +33,23 @@ def __init__(self, args: 'argparse.Namespace', ctrl_addr: str):
...
 
+    @property
+    def _host_ctrl(self) -> str:
+        """
+        Checks if caller (JinaD) has set `docker_kwargs['extra_hosts']` to _docker_host.
+        If yes, set host_ctrl to _docker_host, else set it to localhost
+
+        :return: host for control port
+        """
+        _docker_host = 'host.docker.internal'
```
Comment (maximilianwerk) [inline jina/peapods/runtimes/container/__init__.py:44]: I wonder if this should be some constant somewhere in the code, that can be used at several places!?

File: jina/peapods/runtimes/jinad/__init__.py [hunk 20]
```diff
@@ -17,26 +19,26 @@ def __init__(self, args: 'argparse.Namespace', ctrl_addr: str):
         # process.
         self.host = args.host
         self.port_expose = args.port_expose
-        self.api = PeaDaemonClient(
+        # TODO: args.timeout_ready is always set to -1 for JinadRuntime so that wait_for_success doesn't fail in Pea,
+        # so it can't be used for Client timeout. Setting `timeout` to `None` would wait forever in `DaemonClient`,
+        # which is not ideal. Setting it to 15 secs for testing.
```
Comment (maximilianwerk) [inline jina/peapods/runtimes/jinad/__init__.py:24]: Should this TODO be fixed before merging or can be left over for later? If it stays, please add a name to it.

**Review by maximilianwerk — COMMENTED** [review]
Core code looks reasonable from my point of view. Some minor remarks.

File: jina/peapods/pods/__init__.py (hunk 11 above)
Comment (JoanFM) [reply jina/peapods/pods/__init__.py:137]: I have a PR where I invalidate this, but still not merged

File: jina/peapods/runtimes/container/__init__.py (hunk 19 above)
Comment (deepankarm) [reply jina/peapods/runtimes/container/__init__.py:44]: It is used as a constant (`__docker_host__`) in `daemon/__init__.py`, we might have to move it to `jina/__init__.py`

File: tests/daemon/unit/api/endpoints/test_logs.py (hunk 7 above)
Comment (deepankarm) [reply tests/daemon/unit/api/endpoints/test_logs.py:60]: This is in tests, helps in debugging.

File: tests/daemon/unit/api/endpoints/partial/test_flow.py (hunk 8 above)
Comment (deepankarm) [reply tests/daemon/unit/api/endpoints/partial/test_flow.py:21]: Same.

File: jina/flow/base.py (hunk 15 above)
Comment (deepankarm) [reply jina/flow/base.py:1576]: According to Han's comment above, we are planning to remove this part of the code in https://github.com/jina-ai/jina/pull/2734

File: jina/enums.py (hunk 9 above)
Comment (deepankarm) [reply jina/enums.py:432]: This gets stored in `workspace_store` and gets sent in response as a "string". Using `BetterEnum` would store it as an int, which might not make sense to the client.

File: jina/peapods/runtimes/jinad/__init__.py (hunk 20 above)
Comment (deepankarm) [reply jina/peapods/runtimes/jinad/__init__.py]: changed it to NOTE, as it doesn't need any action now :D

File: tests/unit/peapods/pods/test_networking.py [hunk 21]
```diff
@@ -0,0 +1,92 @@
...
+from jina import __default_host__, __docker_host__
+from jina.peapods.networking import get_connect_host
+from jina.parsers import set_pea_parser
+
+import pytest
+
+docker_uses = 'docker://abc'
+
+
+@pytest.mark.parametrize(
```
Comment (deepankarm) [inline tests/unit/peapods/pods/test_networking.py:11]: @maximilianwerk @JoanFM can you please check if this makes sense?

File: jina/peapods/peas/__init__.py (hunk 16 above)
Comment (deepankarm) [reply jina/peapods/peas/__init__.py:312]: We'll change it in a different PR

**Comment by deepankarm** [comment]
Next tasks https://github.com/jina-ai/jina/issues/2740 & https://github.com/jina-ai/jina/issues/2741

**Review by hanxiao — APPROVED** [review]

File: daemon/api/dependencies.py [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+            # when id doesn't exist in store, create it.
+            store.add(id=self.id, value=WorkspaceState.PENDING)
+            __task_queue__.put((self.id, self.files))
+
+        if self.id in store and store[self.id].state == WorkspaceState.ACTIVE:
+            # when id exists in the store and is "active", update it.
+            store.update(id=self.id, value=WorkspaceState.PENDING)
+            __task_queue__.put((self.id, self.files))
+
+        self.j = {self.id: store[self.id]}
```
Comment (JoanFM) [inline daemon/api/dependencies.py:210]: for next iteration, let's have better naming

**Pull request merged**